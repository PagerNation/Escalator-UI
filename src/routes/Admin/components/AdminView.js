import React from "react";
import "./AdminView.scss";
import { Header, Checkbox, Grid, Button } from 'semantic-ui-react';
import Select from 'react-select';

class AdminView extends React.Component {

  constructor() {
    super();
    this.state = {
      searchValue: "",
      selectedUser: null,
      loading: false,
      sysAdminValue: null,
      valueChanged: false
    };
    _.bindAll(this,
      "handleSearch",
      "getOptions",
      "handleSelectUser",
      "handleChangeSysAdmin",
      "handleUpdateUser"
    );
  }

  handleSelectUser(selection) {
    this.setState({
      selectedUser: selection,
      sysAdminValue: selection && selection.value && selection.value.isSysAdmin,
      valueChanged: false
    });
  }

  handleSearch(value) {
    this.setState({
      searchValue: value,
      loading: true
    });
    if (value !== "") {
      this.props.searchByName(value).then(() => {
        this.setState({
          loading: false
        });
      });
    } else {
      this.setState({
        loading: false
      });
    }
  }

  handleChangeSysAdmin(event, checked) {
    this.setState({
      sysAdminValue: checked.checked,
      valueChanged: true
    });
  }

  handleUpdateUser() {
    this.props.updateUserAdmin(this.state.selectedUser.value._id,
      {isSysAdmin: this.state.sysAdminValue});
    this.setState({
      selectedUser: null
    });
  }

  getOptions() {
    return this.props.searchResults.map((user) => {
      return {
        value: user,
        label: user.name
      }
    });
  }

  renderUserEditor() {
    if(this.state.selectedUser && this.state.selectedUser.value) {
      const user = this.state.selectedUser.value;
      return (
        <div>
          <Header as="h3">
            {user.name}
            <Header.Subheader>
              {user.email}
            </Header.Subheader>
          </Header>
          <Header as="h4">Permissions:</Header>
          <Checkbox
            label='Administrator'
            checked={this.state.sysAdminValue}
            onChange={this.handleChangeSysAdmin}
          />
          <Button
            className="update-btn"
            disabled={!this.state.valueChanged}
            color="green"
            onClick={this.handleUpdateUser}>
            Update
          </Button>
        </div>
      );
    } else {
      return <div></div>;
    }
  }

  render() {
    return (
      <div>
        <Header as="h1">Administration</Header>
        <Header as="h2">Edit User Permissions</Header>
        <Grid>
          <Grid.Column mobile={16} computer={8}>
            <Select
              name="user-search"
              multi={false}
              placeholder="Search users..."
              isLoading={this.state.loading}
              value={this.state.selectedUser}
              options={this.getOptions()}
              onInputChange={this.handleSearch}
              onChange={this.handleSelectUser}
            />
          </Grid.Column>
          <Grid.Column mobile={16} computer={8}>
            {this.renderUserEditor()}
          </Grid.Column>
        </Grid>
      </div>
    );
  }
}

export default AdminView;
