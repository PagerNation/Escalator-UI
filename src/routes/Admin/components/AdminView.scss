.ui.input > div {
  display: block !important;
  width: 100%;

  input {
    width: 100%;
  }
}

.menuItem {
  padding: 5px;
  text-align: left;
}

.menuItem.highlighted {
  background: #eee;
}
