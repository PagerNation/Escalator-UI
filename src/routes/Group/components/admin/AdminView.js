import React from 'react'
import { Icon, Divider, Header, Grid, Segment, Button } from 'semantic-ui-react'
import _ from 'lodash'
import PropTypes from 'prop-types'

class AdminView extends React.Component {
  constructor () {
    super()
    _.bindAll(this,
      'handleProcessRequest',
      'handleAddAdmin')
  }

  handleProcessRequest (userId, approved) {
    this.props.processRequest(this.props.group.name, userId, approved)
  }

  handleAddAdmin (userId) {
    this.props.addAdmin(this.props.group.name, userId)
  }

  handleDowngradeUser (userId) {
    this.props.deleteAdmin(this.props.group.name, userId)
  }

  renderJoinRequests () {
    if (this.props.group.joinRequests && this.props.group.joinRequests.length > 0) {
      return (<Grid className='group-requests'>
        <Grid.Column mobile={16} computer={8}>
          <Header as='h4'>Pending membership requests:</Header>
          <div>
            {this.props.group.joinRequests.map((user, i) =>
              <Segment key={i} raised>
                <Header as='h5'>{user.name}</Header>
                <span>{user.email}</span>
                <Button
                  style={{ marginTop: '-20px' }}
                  floated='right'
                  color='green'
                  onClick={() => this.handleProcessRequest(user._id, true)}>
                    Approve
                  </Button>
                <Button
                  style={{ marginTop: '-20px' }}
                  className='section-btn'
                  floated='right'
                  color='red'
                  onClick={() => this.handleProcessRequest(user._id, false)}>
                    Deny
                  </Button>
              </Segment>
              )}
          </div>
        </Grid.Column>
      </Grid>)
    } else {
      return (<b>No pending join requests</b>)
    }
  }

  userUpgradeButtons (upgrade) {
    const display = {
      'upgrade': {
        icon: 'arrow up',
        color: 'green',
        compare (user, admins) { return admins.indexOf(user._id) === -1 },
        changefunc (comp, user) { return comp.handleAddAdmin(user._id) }
      },
      'downgrade': {
        icon: 'arrow down',
        color: 'red',
        compare (user, admins) { return admins.indexOf(user._id) !== -1 },
        changefunc (comp, user) { return comp.handleDowngradeUser(user._id) }
      }
    }
    return [...this.props.group.users].filter((user) =>
        display[upgrade].compare(user, this.props.group.admins, this.props.user._id)
    ).map((user, i) =>
      <Segment key={i} raised>
        <span as='h5'>{user.name}</span>
        <Button icon
          size='mini'
          style={{ marginTop: '-5px' }}
          floated='right'
          color={display[upgrade].color}
          disabled={user._id === this.props.user._id}
          onClick={() => display[upgrade].changefunc(this, user)}>
          <Icon name={display[upgrade].icon} />
        </Button>
      </Segment>
    )
  }

  renderUpgradeTable () {
    return (<Grid columns={2} divided className='user-upgrade'>
      <Grid.Column mobile={16} computer={8}>
        <Header as='h4'>Members:</Header>
        <div>
          <Segment.Group>
            {this.userUpgradeButtons('upgrade')}
          </Segment.Group>
        </div>
      </Grid.Column>
      <Grid.Column mobile={16} computer={8}>
        <Header as='h4'>Admins:</Header>
        <Segment.Group>
          {this.userUpgradeButtons('downgrade')}
        </Segment.Group>
      </Grid.Column>
    </Grid>)
  }

  render () {
    if (this.props.group.admins.includes(this.props.user._id)) {
      return (<div>
        <Divider />
        <Header as='h3'>Administration</Header>
        {this.renderJoinRequests()}
        <Divider hidden />
        {this.renderUpgradeTable()}
        <Divider />
      </div>)
    } else {
      return null
    }
  }

}

AdminView.propTypes = {
  processRequest: PropTypes.func,
  group: PropTypes.object,
  user: PropTypes.object,
  addAdmin: PropTypes.func,
  deleteAdmin: PropTypes.func
}

export default AdminView
