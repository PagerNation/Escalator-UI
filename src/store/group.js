import 'whatwg-fetch';
import _ from 'lodash';
import { getJSON, deleteObject, putJSON } from "../utils/apiRequest";

// ------------------------------------
// Constants
// ------------------------------------
export const FETCH_GROUP_SUCCESS = 'FETCH_GROUP_SUCCESS';
export const LEAVE_GROUP_SUCCESS = 'LEAVE_GROUP_SUCCESS';
export const UPDATE_EP_SUCCESS = 'UPDATE_EP_SUCCESS';

// ------------------------------------
// Actions
// ------------------------------------
export const fetchGroup = (groupId) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      getJSON('group/' + groupId).then((response) => {
        dispatch({
          type: FETCH_GROUP_SUCCESS,
          payload: response
        });
        resolve();
      });
    });
  }
};

export const leaveGroup = (groupName, userId) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      deleteObject(`group/${groupName}/user/${userId}`).then((json) => {
        dispatch({
          type: LEAVE_GROUP_SUCCESS,
          payload: json
        });
        resolve();
      });
    })
  }
};

export const updateEscalationPolicy = (groupName, ep) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      putJSON(`group/${groupName}/escalationpolicy`, ep).then((response) => {
        dispatch({
          type: UPDATE_EP_SUCCESS,
          payload: response
        });
        resolve();
      });
    })
  }
};


// ------------------------------------
// Action Handlers
// ------------------------------------
const ACTION_HANDLERS = {
  [FETCH_GROUP_SUCCESS]: (state, action) => {
    return action.payload;
  },
  [LEAVE_GROUP_SUCCESS]: (state, action) => {
    return action.payload;
  },
  [UPDATE_EP_SUCCESS]: (state, action) => {
    return action.payload;
  }
};

// ------------------------------------
// Reducer
// ------------------------------------
const initialState = null;

export default function groupReducer(state = initialState, action) {
  const handler = ACTION_HANDLERS[action.type];
  return handler ? handler(state, action) : state;
};
