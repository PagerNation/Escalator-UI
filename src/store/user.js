import 'whatwg-fetch';
import _ from 'lodash';
import jwtDecode from 'jwt-decode';
import { getJSON, postJSON, putJSON, deleteObject } from '../utils/apiRequest';
import { LOG_IN_FAILURE, SIGN_UP_FAILURE } from "./api";

// ------------------------------------
// Constants
// ------------------------------------
export const LOG_IN_SUCCESS = 'LOG_IN_SUCCESS';
export const LOG_OUT = 'LOG_OUT';

export const FETCH_USER_SUCCESS = 'FETCH_USER_SUCCESS';
export const ADD_DEVICE_SUCCESS = 'ADD_DEVICE';
export const DELETE_DEVICE_SUCCESS = 'DELETE_DEVICE_SUCCESS';
export const FETCH_GROUPS_SUCCESS = 'FETCH_GROUPS_SUCCESS';
export const UPDATE_PROFILE_SUCCESS = 'UPDATE_PROFILE_SUCCESS';
export const REORDER_DEVICES_SUCCESS = "REORDER_DEVICES_SUCCESS";


// ------------------------------------
// Actions
// ------------------------------------
export const logIn = (email, password) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      postJSON('auth/login', {email, password}).then((response) => {
        dispatch({
          type: LOG_IN_SUCCESS,
          payload: response
        });
        resolve();
      }).catch((response) => {
        dispatch({
          type: LOG_IN_FAILURE,
          payload: response
        });
        resolve();
      });
    });
  }
};

export const signUp = (name, email, password, passwordConfirm) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      if (password !== passwordConfirm) {
        dispatch({
          type: SIGN_UP_FAILURE,
          payload: "Passwords do not match."
        });
      } else {
        postJSON('auth/signup', {name, email, password}).then((response) => {
          dispatch({
            type: LOG_IN_SUCCESS,
            payload: response
          });
          resolve();
        }).catch((response) => {
          dispatch({
            type: SIGN_UP_FAILURE,
            payload: response
          });
          resolve();
        });
      }
    });
  }
};

export const logOut = () => {
  return (dispatch, getState) => {
    dispatch({
      type: LOG_OUT
    });
  }
};

export const fetchUser = () => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      const userId = jwtDecode(localStorage.getItem('escalatorToken')).id;
      getJSON(`user/${userId}/`).then((response) => {
        dispatch({
          type: FETCH_USER_SUCCESS,
          payload: response
        });
        resolve();
      });
    });
  }
};

export const updateProfile = (profile) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      const userId = jwtDecode(localStorage.getItem('escalatorToken')).id;
      putJSON(`user/${userId}/`, profile).then((response) => {
        dispatch({
          type: UPDATE_PROFILE_SUCCESS,
          payload: response
        });
        resolve();
      });
    });
  }
};

export const addDevice = (device) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      const userId = jwtDecode(localStorage.getItem('escalatorToken')).id;
      postJSON(`user/${userId}/device`, device).then((response) => {
        dispatch({
          type: ADD_DEVICE_SUCCESS,
          payload: response
        });
        resolve();
      });
    })
  }
};

export const deleteDevice = (id) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      const userId = jwtDecode(localStorage.getItem('escalatorToken')).id;
      deleteObject(`user/${userId}/device/${id}`).then((json) => {
        dispatch({
          type: DELETE_DEVICE_SUCCESS,
          payload: json
        });
        resolve();
      });
    })
  }
};

export const reorderDevices = (ids) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      const userId = jwtDecode(localStorage.getItem('escalatorToken')).id;
      putJSON(`user/${userId}/devices/`, {sortOrder: ids}).then((response) => {
        dispatch({
          type: REORDER_DEVICES_SUCCESS,
          payload: response
        });
        resolve();
      });
    })
  }
};

// ------------------------------------
// Action Handlers
// ------------------------------------
const ACTION_HANDLERS = {
  [LOG_IN_SUCCESS]: (state, action) => {
    localStorage.setItem('escalatorToken', action.payload.token);
    return action.payload.user;
  },
  [LOG_OUT]: (state, action) => {
    localStorage.removeItem('escalatorToken');
    return null;
  },
  [FETCH_USER_SUCCESS]: (state, action) => {
    return action.payload;
  },
  [ADD_DEVICE_SUCCESS]: (state, action) => {
    return action.payload;
  },
  [DELETE_DEVICE_SUCCESS]: (state, action) => {
    return action.payload;
  },
  [REORDER_DEVICES_SUCCESS]: (state, action) => {
    return action.payload;
  },
  [FETCH_GROUPS_SUCCESS]: (state, action) => {
    const newState = _.extend({}, state);
    newState.groups = action.payload.groups;
    return newState;
  },
  [UPDATE_PROFILE_SUCCESS]: (state, action) => {
    return action.payload;
  }
};

// ------------------------------------
// Reducer
// ------------------------------------
const initialState = null;

export default function userReducer(state = initialState, action) {
  const handler = ACTION_HANDLERS[action.type];
  return handler ? handler(state, action) : state;
};
