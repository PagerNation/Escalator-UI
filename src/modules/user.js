import { injectReducer } from '../store/reducers';

// ------------------------------------
// Constants
// ------------------------------------
export const FETCH_USER = 'FETCH_USER';

// ------------------------------------
// Actions
// ------------------------------------
export function fetchUser() {
  return {
    type: FETCH_USER,
    payload: value
  }
}

export const actions = {
  fetchUser
};

// ------------------------------------
// Action Handlers
// ------------------------------------
const ACTION_HANDLERS = {
  [FETCH_USER]: (state, action) => {
    // TODO: Use endpoint
    return {
      id: '693d9a0698aff95c',
      name: 'Kevin Moses',
      email: 'kwm4385@rit.edu',
      groups: [{
        type: "Group",
        ref: 'Some group'
      },
        {
          type: "Group",
          ref: 'Awesome Group'
        }],
      devices: [{
        name: 'Cell',
        type: 'sms',
        contactInformation: '585-328-5354',
        createdAt: Date.now()
      }]
    };
  }
};

// ------------------------------------
// Reducer
// ------------------------------------
const initialState = null;

export default function userReducer(state = initialState, action) {
  const handler = ACTION_HANDLERS[action.type];

  return handler ? handler(state, action) : state;
};

injectReducer(store, {
  key: 'user',
  reducer: userReducer
});
