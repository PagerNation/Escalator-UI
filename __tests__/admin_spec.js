import React from 'react';
import AdminView from '../src/routes/Group/components/admin/AdminView';
import renderer from 'react-test-renderer';
import timezone_mock from 'timezone-mock';

import group from './fixtures/group_response.json';
import user from './fixtures/user.json';
import tickets from './fixtures/tickets.json'

describe('AdminView', () => {
  it('renders AdminView', () => {
    timezone_mock.register('US/Eastern');
    group.admins.push(user._id);
    const component = renderer.create(<AdminView group={group} user={user} />);
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  })
});

