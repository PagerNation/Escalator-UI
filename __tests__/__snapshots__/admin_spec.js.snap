// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminView renders AdminView 1`] = `
<div>
  <div
    className="ui divider"
  />
  <h3
    className="ui header"
  >
    Administration
  </h3>
  <b>
    No pending join requests
  </b>
  <div
    className="ui hidden divider"
  />
  <div
    className="ui divided two column grid user-upgrade"
  >
    <div
      className="eight wide computer sixteen wide mobile column"
    >
      <h4
        className="ui header"
      >
        Members:
      </h4>
      <div>
        <div
          className="ui segments"
          hidden={false}
        >
          <div
            className="ui raised segment"
          >
            <span
              as="h5"
            >
              Demetrius.Hahn
            </span>
            <button
              className="ui green mini right floated icon button"
              onClick={[Function]}
              style={
                Object {
                  "marginTop": "-5px",
                }
              }
              tabIndex={undefined}
            >
              <i
                className="arrow up icon"
              />
            </button>
          </div>
          <div
            className="ui raised segment"
          >
            <span
              as="h5"
            >
              Seth4
            </span>
            <button
              className="ui green mini right floated icon button"
              onClick={[Function]}
              style={
                Object {
                  "marginTop": "-5px",
                }
              }
              tabIndex={undefined}
            >
              <i
                className="arrow up icon"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="eight wide computer sixteen wide mobile column"
    >
      <h4
        className="ui header"
      >
        Admins:
      </h4>
      <div
        className="ui segments"
      >
        <div
          className="ui raised segment"
        >
          <span
            as="h5"
          >
            Kelly85
          </span>
          <button
            className="ui red mini disabled right floated icon button"
            onClick={[Function]}
            style={
              Object {
                "marginTop": "-5px",
              }
            }
            tabIndex={undefined}
          >
            <i
              className="arrow down icon"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="ui divider"
  />
</div>
`;
